File: desktop-mend/src/test/java/helper/CryptoHelperTest.java
package helper;

import co.samco.mend4.core.AppProperties;
import co.samco.mend4.core.Settings;
import co.samco.mend4.core.exception.CorruptSettingsException;
import co.samco.mend4.core.exception.MalformedLogFileException;
import co.samco.mend4.core.util.LogUtils;
import co.samco.mend4.desktop.helper.CryptoHelper;
import commands.TestBase;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.mockito.ArgumentCaptor;

import javax.crypto.BadPaddingException;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.NoSuchPaddingException;
import java.io.File;
import java.io.IOException;
import java.security.InvalidAlgorithmParameterException;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.security.spec.InvalidKeySpecException;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import static org.mockito.Matchers.any;
import static org.mockito.Matchers.eq;
import static org.mockito.Mockito.*;

public class CryptoHelperTest extends TestBase {

    private String encdir = File.separatorChar + "output" + File.separatorChar + "directory";

    @Before
    public void setup() {
        super.setup();
        cryptoHelper = new CryptoHelper(strings, log, fileResolveHelper, settings, cryptoProvider,
                keyHelper, osDao, versionHelper);
    }

    @Test
    public void testEncInputOutputFilesCorrect() throws IOException, NoSuchPaddingException, NoSuchAlgorithmException,
            InvalidKeyException, InvalidAlgorithmParameterException, IllegalBlockSizeException, BadPaddingException,
            CorruptSettingsException, InvalidKeySpecException {
        String inputFileName = "/input.txt";
        File inputFile = new File(inputFileName);
        when(settings.getValue(Settings.Name.ENCDIR)).thenReturn(encdir);
        cryptoHelper.encryptFile(inputFile, null);

        ArgumentCaptor<File> fileCaptor1 = ArgumentCaptor.forClass(File.class);
        verify(osDao).getInputStreamForFile(fileCaptor1.capture());
        Assert.assertEquals(inputFileName, fileCaptor1.getValue().getAbsolutePath());

        ArgumentCaptor<File> fileCaptor2 = ArgumentCaptor.forClass(File.class);
        verify(osDao).getOutputStreamForFile(fileCaptor2.capture());
        Pattern pattern = Pattern.compile(encdir + File.separatorChar + "\\d{17}." + AppProperties.ENC_FILE_EXTENSION);
        Matcher matcher = pattern.matcher(fileCaptor2.getValue().getAbsolutePath());
        Assert.assertTrue(matcher.matches());

        verify(cryptoProvider).encryptEncStream(eq(null), any(), any(), eq("txt"));
    }

    @Test
    public void testEncInputFileNoExtension() throws IOException, NoSuchPaddingException, NoSuchAlgorithmException,
            InvalidKeyException, InvalidAlgorithmParameterException, IllegalBlockSizeException, BadPaddingException,
            CorruptSettingsException, InvalidKeySpecException {
        String inputFileName = "/input";
        File inputFile = new File(inputFileName);
        when(settings.getValue(Settings.Name.ENCDIR)).thenReturn(encdir);
        cryptoHelper.encryptFile(inputFile, null);
        ArgumentCaptor<File> fileCaptor = ArgumentCaptor.forClass(File.class);
        verify(osDao).getInputStreamForFile(fileCaptor.capture());
        Assert.assertEquals(inputFileName, fileCaptor.getValue().getAbsolutePath());
    }

    @Test(expected = IllegalArgumentException.class)
    public void testDontOverwriteEncFile() throws IOException, NoSuchPaddingException, NoSuchAlgorithmException,
            InvalidKeyException, InvalidAlgorithmParameterException, IllegalBlockSizeException, BadPaddingException,
            CorruptSettingsException, InvalidKeySpecException {
        File inputFile = new File("input");
        when(settings.getValue(Settings.Name.ENCDIR)).thenReturn(encdir);
        doThrow(new IllegalArgumentException()).when(fileResolveHelper).assertFileDoesNotExist(any());
        cryptoHelper.encryptFile(inputFile, null);
    }

    @Test
    public void testEncEmptyLog() throws IOException, NoSuchPaddingException, NoSuchAlgorithmException, InvalidKeyException,
            InvalidAlgorithmParameterException, IllegalBlockSizeException, BadPaddingException, CorruptSettingsException,
            InvalidKeySpecException {
        cryptoHelper.encryptTextToLog(new char[0], true);
        verify(fileResolveHelper, never()).getCurrentLogFile();
        verify(cryptoProvider, never()).encryptLogStream(any(), any(), any());
    }

    @Test
    public void testLogEncrypted() throws IOException, NoSuchPaddingException, NoSuchAlgorithmException, InvalidKeyException,
            InvalidAlgorithmParameterException, IllegalBlockSizeException, BadPaddingException, CorruptSettingsException,
            InvalidKeySpecException {
        String message = "hi";
        File logFile = new File("/currentLogFile." + AppProperties.LOG_FILE_EXTENSION);
        when(fileResolveHelper.getCurrentLogFile()).thenReturn(logFile);
        cryptoHelper.encryptTextToLog(message.toCharArray(), true);
        verify(osDao).createNewFile(any());
        verify(osDao).getOutputStreamForFile(eq(logFile), eq(true));
        verify(cryptoProvider).encryptLogStream(any(), eq(message), any());
    }

    @Test
    public void testHeaderAddedToLog() throws IOException, NoSuchPaddingException, NoSuchAlgorithmException, InvalidKeyException,
            InvalidAlgorithmParameterException, IllegalBlockSizeException, BadPaddingException, CorruptSettingsException,
            InvalidKeySpecException {
        String version = "ver";
        String message = "hi";
        String messageWithHeader = LogUtils.addHeaderToLogText(message, strings.get("Platform.header"), version, "\n");
        File logFile = new File("/currentLogFile." + AppProperties.LOG_FILE_EXTENSION);
        when(versionHelper.getVersion()).thenReturn(version);
        when(fileResolveHelper.getCurrentLogFile()).thenReturn(logFile);
        cryptoHelper.encryptTextToLog(message.toCharArray(), false);
        verify(cryptoProvider).encryptLogStream(any(), eq(messageWithHeader), any());
    }

    @Test
    public void testDecryptLog() throws IOException, NoSuchPaddingException, NoSuchAlgorithmException, InvalidKeyException,
            InvalidAlgorithmParameterException, IllegalBlockSizeException, BadPaddingException, MalformedLogFileException,
            InvalidKeySpecException {
        File logFile = new File("/logfile.log");
        cryptoHelper.decryptLog(logFile);
        verify(osDao).getInputStreamForFile(eq(logFile));
        verify(cryptoProvider).decryptLogStream(any(), any(), eq(out));
    }

    @Test
    public void testDecryptFile() throws IOException, NoSuchPaddingException, NoSuchAlgorithmException, InvalidKeyException,
            InvalidAlgorithmParameterException, MalformedLogFileException, BadPaddingException, CorruptSettingsException,
            InvalidKeySpecException, IllegalBlockSizeException {
        String encFileName = "encFile";
        String fileExtension = "txt";
        File encFile = new File(File.separator + "path" + File.separator + encFileName + "." + AppProperties.ENC_FILE_EXTENSION);
        ArgumentCaptor<File> fileCaptor = ArgumentCaptor.forClass(File.class);
        String decDir = File.separator + "decdir";
        when(settings.getValue(Settings.Name.DECDIR)).thenReturn(decDir);
        when(cryptoProvider.decryptEncStream(any(), any(), any())).thenReturn(fileExtension);
        cryptoHelper.decryptFile(encFile, false);
        verify(fileResolveHelper).assertDirWritable(decDir);
        verify(fileResolveHelper, times(2)).assertFileDoesNotExist(fileCaptor.capture());
        Assert.assertEquals(fileCaptor.getAllValues().get(0).getAbsolutePath(),
                decDir + File.separatorChar + encFileName);
        Assert.assertEquals(fileCaptor.getAllValues().get(1).getAbsolutePath(),
                decDir + File.separatorChar + encFileName + "." + fileExtension);
        verify(cryptoProvider).decryptEncStream(any(), any(), any());
        verify(osDao).renameFile(any(), eq(encFileName + "." + fileExtension));
        verify(osDao).desktopOpenFile(any());
    }
}

File: desktop-mend/src/test/java/helper/CryptoHelperTest.kt
package helper

import co.samco.mend4.core.AppProperties
import co.samco.mend4.core.util.LogUtils
import co.samco.mend4.desktop.dao.SettingsDao
import co.samco.mend4.desktop.exception.MendLockedException
import co.samco.mend4.desktop.helper.impl.CryptoHelperImpl
import commands.TestBase
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.argumentCaptor
import com.nhaarman.mockitokotlin2.doThrow
import com.nhaarman.mockitokotlin2.eq
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.never
import com.nhaarman.mockitokotlin2.times
import com.nhaarman.mockitokotlin2.verify
import com.nhaarman.mockitokotlin2.whenever
import testutils.TestUtils
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.File
import java.security.PrivateKey
import java.util.regex.Pattern

class CryptoHelperTest : TestBase() {
    private val encDir = File.separatorChar.toString() + "output" + File.separatorChar + "directory"
    private var privateKey: PrivateKey = mock()

    @Before
    override fun setup() {
        super.setup()
        cryptoHelper = CryptoHelperImpl(
            strings = strings,
            log = log,
            fileResolveHelper = fileResolveHelper,
            settings = settings,
            cryptoProvider = cryptoProvider,
            keyHelper = keyHelper,
            versionHelper = versionHelper,
            osDao = osDao
        )
    }

    @Test
    fun testEncInputOutputFilesCorrect() {
        val inputFileName = "/input.txt"
        val inputFile = File(inputFileName)
        whenever(settings.getValue(SettingsDao.ENC_DIR)).thenReturn(encDir)
        whenever(osDao.fileInputStream(any())).thenReturn(TestUtils.emptyInputStream)
        whenever(osDao.fileOutputSteam(any(), any())).thenReturn(ByteArrayOutputStream())

        cryptoHelper.encryptFile(inputFile, null)

        val fileCaptor1 = argumentCaptor<File>()
        verify(osDao).fileInputStream(fileCaptor1.capture())
        assertEquals(inputFileName, fileCaptor1.firstValue.absolutePath)

        val fileCaptor2 = argumentCaptor<File>()
        verify(osDao).fileOutputSteam(fileCaptor2.capture(), eq(false))
        val pattern =
            Pattern.compile(encDir + File.separatorChar + "\\d{17}." + AppProperties.ENC_FILE_EXTENSION)
        val matcher = pattern.matcher(fileCaptor2.firstValue.absolutePath)
        assertTrue(matcher.matches())
        verify(cryptoProvider).encryptEncStream(any(), any(), eq("txt"))
    }

    @Test
    fun testEncInputFileNoExtension() {
        val inputFileName = "/input"
        val inputFile = File(inputFileName)
        whenever(settings.getValue(SettingsDao.ENC_DIR)).thenReturn(encDir)
        cryptoHelper.encryptFile(inputFile, null)
        val fileCaptor = argumentCaptor<File>()
        verify(osDao).fileInputStream(fileCaptor.capture())
        assertEquals(inputFileName, fileCaptor.firstValue.absolutePath)
    }

    @Test(expected = IllegalArgumentException::class)
    fun testDontOverwriteEncFile() {
        val inputFile = File("input")
        whenever(settings.getValue(SettingsDao.ENC_DIR)).thenReturn(encDir)
        doThrow(IllegalArgumentException())
            .whenever(fileResolveHelper)
            .assertFileDoesNotExist(any())
        cryptoHelper.encryptFile(inputFile, null)
    }

    @Test
    fun testEncEmptyLog() {
        cryptoHelper.encryptTextToLog(CharArray(0), true)
        verify(fileResolveHelper, never()).currentLogFile
        verify(cryptoProvider, never()).encryptLogStream(any(), any())
    }

    @Test
    fun testLogEncrypted() {
        val message = "hi"
        val logFile = File("/currentLogFile." + AppProperties.LOG_FILE_EXTENSION)
        whenever(fileResolveHelper.currentLogFile).thenReturn(logFile)
        whenever(osDao.fileOutputSteam(eq(logFile), eq(true))).thenReturn(ByteArrayOutputStream())

        cryptoHelper.encryptTextToLog(message.toCharArray(), true)

        verify(osDao).createNewFile(any())
        verify(osDao).fileOutputSteam(eq(logFile), eq(true))
        verify(cryptoProvider).encryptLogStream(eq(message), any())
    }

    @Test
    fun testHeaderAddedToLog() {
        val version = "ver"
        val message = "hi"
        val messageWithHeader: String =
            LogUtils.addHeaderToLogText(message, strings["Platform.header"], version, "\n")
        val logFile = File("/currentLogFile." + AppProperties.LOG_FILE_EXTENSION)
        whenever(versionHelper.version).thenReturn(version)
        whenever(fileResolveHelper.currentLogFile).thenReturn(logFile)
        whenever(osDao.fileOutputSteam(eq(logFile), eq(true))).thenReturn(ByteArrayOutputStream())

        cryptoHelper.encryptTextToLog(message.toCharArray(), false)

        verify(cryptoProvider).encryptLogStream(eq(messageWithHeader), any())
    }

    @Test
    fun testDecryptLog() {
        val logFile = File("/logfile.log")
        whenever(keyHelper.privateKey).thenReturn(privateKey)
        whenever(osDao.fileInputStream(eq(logFile))).thenReturn(TestUtils.emptyInputStream)

        cryptoHelper.decryptLog(logFile)

        verify(osDao).fileInputStream(eq(logFile))
        verify(cryptoProvider).decryptLogStream(any(), any(), eq(out))
    }

    @Test(expected = MendLockedException::class)
    fun testDecryptLogMendLocked() {
        cryptoHelper.decryptLog(File(""))
    }

    @Test
    fun testDecryptFile() {
        val encFileName = "encFile"
        val fileExtension = "txt"
        val encFile =
            File(File.separator + "path" + File.separator + encFileName + "." + AppProperties.ENC_FILE_EXTENSION)
        val fileCaptor = argumentCaptor<File>()
        val decDir = File.separator + "decDir"
        whenever(cryptoProvider.decryptEncStream(any(), any(), any())).thenReturn(fileExtension)
        whenever(keyHelper.privateKey).thenReturn(privateKey)
        whenever(osDao.fileOutputSteam(any(), any())).thenReturn(ByteArrayOutputStream())
        whenever(osDao.fileInputStream(any())).thenReturn(TestUtils.emptyInputStream)

        cryptoHelper.decryptFile(encFile, decDir, false)

        verify(fileResolveHelper).assertDirWritable(decDir)
        verify(fileResolveHelper, times(2)).assertFileDoesNotExist(fileCaptor.capture())
        assertEquals(
            decDir + File.separatorChar + encFileName,
            fileCaptor.allValues[0].absolutePath,
        )
        assertEquals(
            decDir + File.separatorChar + encFileName + "." + fileExtension,
            fileCaptor.allValues[1].absolutePath
        )
        verify(cryptoProvider).decryptEncStream(any(), any(), any())
        verify(osDao).renameFile(
            any(),
            eq(File(decDir + File.separator + encFileName + "." + fileExtension))
        )
        verify(osDao).desktopOpenFile(any())
    }

    @Test(expected = MendLockedException::class)
    fun testDecryptFileMendLocked() {
        cryptoHelper.decryptFile(File(""), "", false)
    }
}